File: WPF/Controls/Popup.vb
Imports Ks.Common.MVVM

Namespace Controls

    Public Class Popup
        Inherits ContentElement

        Shared Sub New()
            DefaultStyleKeyProperty.OverrideMetadata(GetType(Popup), New FrameworkPropertyMetadata(GetType(Popup)))
            HorizontalAlignmentProperty.OverrideMetadata(GetType(Popup), New FrameworkPropertyMetadata(HorizontalAlignment.Center))
            VerticalAlignmentProperty.OverrideMetadata(GetType(Popup), New FrameworkPropertyMetadata(VerticalAlignment.Center))
            'WidthProperty.OverrideMetadata(GetType(Popup), New FrameworkPropertyMetadata(300.0))
            'HeightProperty.OverrideMetadata(GetType(Popup), New FrameworkPropertyMetadata(300.0))
        End Sub

#Region "BeforeShow Event"
        Public Shared ReadOnly BeforeShowEvent As RoutedEvent = EventManager.RegisterRoutedEvent("BeforeShow", RoutingStrategy.Direct, GetType(RoutedEventHandler), GetType(Popup))

        Protected Overridable Sub OnBeforeShow()
            Dim E = New RoutedEventArgs(BeforeShowEvent)
            Me.RaiseEvent(E)
        End Sub

        Public Custom Event BeforeShow As RoutedEventHandler
            AddHandler(ByVal value As RoutedEventHandler)
                Me.AddHandler(BeforeShowEvent, value)
            End AddHandler
            RemoveHandler(ByVal value As RoutedEventHandler)
                Me.RemoveHandler(BeforeShowEvent, value)
            End RemoveHandler
            RaiseEvent()
            End RaiseEvent
        End Event
#End Region

#Region "Shown Event"
        Public Shared ReadOnly ShownEvent As RoutedEvent = EventManager.RegisterRoutedEvent("Shown", RoutingStrategy.Direct, GetType(RoutedEventHandler), GetType(Popup))

        Protected Overridable Sub OnShown()
            Dim E = New RoutedEventArgs(ShownEvent)
            Me.RaiseEvent(E)
        End Sub

        Public Custom Event Shown As RoutedEventHandler
            AddHandler(ByVal value As RoutedEventHandler)
                Me.AddHandler(ShownEvent, value)
            End AddHandler
            RemoveHandler(ByVal value As RoutedEventHandler)
                Me.RemoveHandler(ShownEvent, value)
            End RemoveHandler
            RaiseEvent()
            End RaiseEvent
        End Event
#End Region

#Region "Hid Event"
        Public Shared ReadOnly HidEvent As RoutedEvent = EventManager.RegisterRoutedEvent("Hid", RoutingStrategy.Direct, GetType(RoutedEventHandler), GetType(Popup))

        Protected Overridable Sub OnHid()
            Dim E = New RoutedEventArgs(HidEvent)
            Me.RaiseEvent(E)
        End Sub

        Public Custom Event Hid As RoutedEventHandler
            AddHandler(ByVal value As RoutedEventHandler)
                Me.AddHandler(HidEvent, value)
            End AddHandler
            RemoveHandler(ByVal value As RoutedEventHandler)
                Me.RemoveHandler(HidEvent, value)
            End RemoveHandler
            RaiseEvent()
            End RaiseEvent
        End Event
#End Region

#Region "Show Command"
        Private _ShowCommand As DelegateCommand = New DelegateCommand(AddressOf Me.Show)

        Public Sub Show()
            Me.IsShown = True
        End Sub

        Public ReadOnly Property ShowCommand As DelegateCommand
            Get
                Return Me._ShowCommand
            End Get
        End Property
#End Region

#Region "Hide Command"
        Private _HideCommand As DelegateCommand = New DelegateCommand(AddressOf Me.Hide)

        Public Sub Hide()
            Me.IsShown = False
        End Sub

        Public ReadOnly Property HideCommand As DelegateCommand
            Get
                Return Me._HideCommand
            End Get
        End Property
#End Region

#Region "IsShown Property"
        Public Shared ReadOnly IsShownProperty As DependencyProperty = DependencyProperty.Register("IsShown", GetType(Boolean), GetType(Popup), New PropertyMetadata(False, AddressOf IsShown_Changed))

        Private Shared Sub IsShown_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, Popup)

            If DirectCast(E.NewValue, Boolean) Then
                Self.OnBeforeShow()
                Self.GetWindow()?.AddPopup(Self)
                Self.OnShown()
            Else
                Self.GetWindow()?.RemovePopup(Self)
                Self.OnHid()
            End If
        End Sub

        Public Property IsShown As Boolean
            Get
                Return DirectCast(Me.GetValue(IsShownProperty), Boolean)
            End Get
            Set(ByVal value As Boolean)
                Me.SetValue(IsShownProperty, value)
            End Set
        End Property
#End Region

#Region "HasShelter Property"
        Public Shared ReadOnly HasShelterProperty As DependencyProperty = DependencyProperty.Register("HasShelter", GetType(Boolean), GetType(Popup), New PropertyMetadata(True, AddressOf HasShelter_Changed))

        Private Shared Sub HasShelter_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, Popup)

            'Dim OldValue = DirectCast(E.OldValue, Boolean)
            'Dim NewValue = DirectCast(E.NewValue, Boolean)

            If Self.IsShown Then
                Self.GetWindow()?.RefreshPopup(Self)
            End If
        End Sub

        Public Property HasShelter As Boolean
            Get
                Return DirectCast(Me.GetValue(HasShelterProperty), Boolean)
            End Get
            Set(ByVal value As Boolean)
                Me.SetValue(HasShelterProperty, value)
            End Set
        End Property
#End Region

#Region "DimShelter Property"
        Public Shared ReadOnly DimShelterProperty As DependencyProperty = DependencyProperty.Register("DimShelter", GetType(Boolean), GetType(Popup), New PropertyMetadata(True, AddressOf DimShelter_Changed))

        Private Shared Sub DimShelter_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, Popup)

            'Dim OldValue = DirectCast(E.OldValue, Boolean)
            'Dim NewValue = DirectCast(E.NewValue, Boolean)

            If Self.IsShown Then
                Self.GetWindow()?.UpdateDims()
            End If
        End Sub

        Public Property DimShelter As Boolean
            Get
                Return DirectCast(Me.GetValue(DimShelterProperty), Boolean)
            End Get
            Set(ByVal value As Boolean)
                Me.SetValue(DimShelterProperty, value)
            End Set
        End Property
#End Region

#Region "Layer Property"
        Public Shared ReadOnly LayerProperty As DependencyProperty = DependencyProperty.Register("Layer", GetType(Integer), GetType(Popup), New PropertyMetadata(0, AddressOf Layer_Changed))

        Private Shared Sub Layer_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, Popup)

            If Self.IsShown Then
                Dim OldValue = DirectCast(E.OldValue, Integer)
                'Dim NewValue = DirectCast(E.NewValue, Integer)

                Dim Window = Self.GetWindow()

                Window?.RemovePopup(Self, OldValue)
                Window?.AddPopup(Self)
            End If
        End Sub

        Public Property Layer As Integer
            Get
                Return DirectCast(Me.GetValue(LayerProperty), Integer)
            End Get
            Set(ByVal value As Integer)
                Me.SetValue(LayerProperty, value)
            End Set
        End Property
#End Region

#Region "IsShelterSensitive Property"
        Public Shared ReadOnly IsShelterSensitiveProperty As DependencyProperty = DependencyProperty.Register("IsShelterSensitive", GetType(Boolean), GetType(PopupShelter), New PropertyMetadata(True, AddressOf IsShelterSensitive_Changed))

        Private Shared Sub IsShelterSensitive_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupShelter)

            Dim OldValue = DirectCast(E.OldValue, Boolean)
            Dim NewValue = DirectCast(E.NewValue, Boolean)
        End Sub

        Public Property IsShelterSensitive As Boolean
            Get
                Return DirectCast(Me.GetValue(IsShelterSensitiveProperty), Boolean)
            End Get
            Set(ByVal value As Boolean)
                Me.SetValue(IsShelterSensitiveProperty, value)
            End Set
        End Property
#End Region

#Region "Window Property"
        Public Shared ReadOnly WindowProperty As DependencyProperty = DependencyProperty.Register("Window", GetType(Window), GetType(Popup), New PropertyMetadata(Nothing, AddressOf Window_Changed))

        Private Shared Sub Window_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, Popup)

            Dim OldValue = GetWindow(DirectCast(E.OldValue, Window))
            Dim NewValue = GetWindow(DirectCast(E.NewValue, Window))

            If Self.IsShown Then
                OldValue?.RemovePopup(Self)
                NewValue?.AddPopup(Self)
            End If
        End Sub

        Public Property Window As Window
            Get
                Return DirectCast(Me.GetValue(WindowProperty), Window)
            End Get
            Set(ByVal value As Window)
                Me.SetValue(WindowProperty, value)
            End Set
        End Property

        Private Shared Function GetWindow(ByVal Window As Window) As Window
            If Window IsNot Nothing Then
                Return Window
            End If
            Return TryCast(KsApplication.Current.Window.View, Window)
        End Function

        Private Function GetWindow() As Window
            Return GetWindow(Me.Window)
        End Function
#End Region

        Friend ArrangeCallBack As Func(Of PopupPanel, Size, Size, Rect?)

    End Class

End Namespace

File: UI/WPF/Controls/Popup.cs
using System;
using System.Windows;

using Ks.Common.MVVM;

namespace Ks.Common.Controls
{
    public class Popup : ContentElement
    {
        static Popup()
        {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(Popup), new FrameworkPropertyMetadata(typeof(Popup)));
            HorizontalAlignmentProperty.OverrideMetadata(typeof(Popup), new FrameworkPropertyMetadata(HorizontalAlignment.Center));
            VerticalAlignmentProperty.OverrideMetadata(typeof(Popup), new FrameworkPropertyMetadata(VerticalAlignment.Center));
        }

        public Popup()
        {
            this.ShowCommand = new DelegateCommand(this.Show);
            this.HideCommand = new DelegateCommand(this.Hide);
        }

        public static readonly RoutedEvent BeforeShowEvent = EventManager.RegisterRoutedEvent("BeforeShow", RoutingStrategy.Direct, typeof(RoutedEventHandler), typeof(Popup));

        protected virtual void OnBeforeShow()
        {
            var E = new RoutedEventArgs(BeforeShowEvent);
            this.RaiseEvent(E);
        }

        public event RoutedEventHandler BeforeShow
        {
            add
            {
                this.AddHandler(BeforeShowEvent, value);
            }
            remove
            {
                this.RemoveHandler(BeforeShowEvent, value);
            }
        }

        public static readonly RoutedEvent ShownEvent = EventManager.RegisterRoutedEvent("Shown", RoutingStrategy.Direct, typeof(RoutedEventHandler), typeof(Popup));

        protected virtual void OnShown()
        {
            var E = new RoutedEventArgs(ShownEvent);
            this.RaiseEvent(E);
        }

        public event RoutedEventHandler Shown
        {
            add
            {
                this.AddHandler(ShownEvent, value);
            }
            remove
            {
                this.RemoveHandler(ShownEvent, value);
            }
        }

        public static readonly RoutedEvent HidEvent = EventManager.RegisterRoutedEvent("Hid", RoutingStrategy.Direct, typeof(RoutedEventHandler), typeof(Popup));

        protected virtual void OnHid()
        {
            var E = new RoutedEventArgs(HidEvent);
            this.RaiseEvent(E);
        }

        public event RoutedEventHandler Hid
        {
            add
            {
                this.AddHandler(HidEvent, value);
            }
            remove
            {
                this.RemoveHandler(HidEvent, value);
            }
        }

        public void Show()
        {
            this.IsShown = true;
        }

        public DelegateCommand ShowCommand { get; }

        public void Hide()
        {
            this.IsShown = false;
        }

        public DelegateCommand HideCommand { get; }

        public static readonly DependencyProperty IsShownProperty = DependencyProperty.Register("IsShown", typeof(bool), typeof(Popup), new PropertyMetadata(false, IsShown_Changed));

        private static void IsShown_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (Popup) D;

            if ((bool) E.NewValue)
            {
                Self.OnBeforeShow();
                Self.GetWindow()?.AddPopup(Self);
                Self.OnShown();
            }
            else
            {
                Self.GetWindow()?.RemovePopup(Self);
                Self.OnHid();
            }
        }

        public bool IsShown
        {
            get => (bool) this.GetValue(IsShownProperty);
            set => this.SetValue(IsShownProperty, value);
        }

        public static readonly DependencyProperty HasShelterProperty = DependencyProperty.Register("HasShelter", typeof(bool), typeof(Popup), new PropertyMetadata(true, HasShelter_Changed));

        private static void HasShelter_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (Popup) D;

            // Dim OldValue = DirectCast(E.OldValue, Boolean)
            // Dim NewValue = DirectCast(E.NewValue, Boolean)

            if (Self.IsShown)
            {
                Self.GetWindow()?.RefreshPopup(Self);
            }
        }

        public bool HasShelter
        {
            get => (bool) this.GetValue(HasShelterProperty);
            set => this.SetValue(HasShelterProperty, value);
        }

        public static readonly DependencyProperty DimShelterProperty = DependencyProperty.Register("DimShelter", typeof(bool), typeof(Popup), new PropertyMetadata(true, DimShelter_Changed));

        private static void DimShelter_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (Popup) D;

            // Dim OldValue = DirectCast(E.OldValue, Boolean)
            // Dim NewValue = DirectCast(E.NewValue, Boolean)

            if (Self.IsShown)
            {
                Self.GetWindow()?.UpdateDims();
            }
        }

        public bool DimShelter
        {
            get => (bool) this.GetValue(DimShelterProperty);
            set => this.SetValue(DimShelterProperty, value);
        }

        public static readonly DependencyProperty LayerProperty = DependencyProperty.Register("Layer", typeof(int), typeof(Popup), new PropertyMetadata(0, Layer_Changed));

        private static void Layer_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (Popup) D;

            if (Self.IsShown)
            {
                var OldValue = (int) E.OldValue;
                // Dim NewValue = DirectCast(E.NewValue, Integer)

                var Window = Self.GetWindow();

                Window?.RemovePopup(Self, OldValue);
                Window?.AddPopup(Self);
            }
        }

        public int Layer
        {
            get => (int) this.GetValue(LayerProperty);
            set => this.SetValue(LayerProperty, value);
        }

        public static readonly DependencyProperty IsShelterSensitiveProperty = DependencyProperty.Register("IsShelterSensitive", typeof(bool), typeof(PopupShelter), new PropertyMetadata(true, IsShelterSensitive_Changed));

        private static void IsShelterSensitive_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            // var Self = (PopupShelter) D;

            // var OldValue = (bool) E.OldValue;
            // var NewValue = (bool) E.NewValue;
        }

        public bool IsShelterSensitive
        {
            get => (bool) this.GetValue(IsShelterSensitiveProperty);
            set => this.SetValue(IsShelterSensitiveProperty, value);
        }

        public static readonly DependencyProperty WindowProperty = DependencyProperty.Register("Window", typeof(Window), typeof(Popup), new PropertyMetadata(null, Window_Changed));

        private static void Window_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (Popup) D;

            var OldValue = GetWindow((Window) E.OldValue);
            var NewValue = GetWindow((Window) E.NewValue);

            if (Self.IsShown)
            {
                OldValue?.RemovePopup(Self);
                NewValue?.AddPopup(Self);
            }
        }

        public Window Window
        {
            get => (Window) this.GetValue(WindowProperty);
            set => this.SetValue(WindowProperty, value);
        }

        private static Window GetWindow(Window Window)
        {
            if (Window != null)
            {
                return Window;
            }

            return KsApplication.Current.Window.View as Window;
        }

        private Window GetWindow()
        {
            return GetWindow(this.Window);
        }

        internal Func<PopupPanel, Size, Size, Rect?> ArrangeCallBack;
    }
}
